File: Java-Civitas/Jugador.java
package civitas;

import java.util.ArrayList;

public class Jugador implements Comparable<Jugador> {

    final protected int CasasMax = 4;
    final protected int CasasPorHotel = 4;
    protected boolean encarcelado;
    final protected int HotelesMax = 4;
    private String nombre;
    private int numCasillaActual;
    final protected float PasoPorSalida = 1000;
    final protected float PrecioLibertad = 200;
    private boolean puedeComprar;
    private float saldo;
    final private float SaldoInicial = 7500;
    private ArrayList<TituloPropiedad> propiedades;
    private Sorpresa salvoconducto;

    public Jugador(String nomb) {
        nombre = nomb;
    }

    protected Jugador(Jugador jug) {
        propiedades = new ArrayList<>(jug.propiedades);
    }

    int cantidadCasasHoteles() {
        return propiedades.size();
    }

    @Override
    public int compareTo(Jugador jug) {
        int comp = 0;

        if (getSaldo() > jug.getSaldo()) {
            comp = 1;
        } else if (getSaldo() < jug.getSaldo()) {
            comp = -1;
        }

        return comp;
    }

    boolean comprar(TituloPropiedad propiedad) {
        throw new UnsupportedOperationException("No implementado");
    }

    boolean construirCasa(int ip) {
        throw new UnsupportedOperationException("No implementado");
    }

    boolean construirHotel(int ip) {
        throw new UnsupportedOperationException("No implementado");
    }

    protected boolean debeSerEncarcelado() {
        boolean encarcelar = false;

        if (!isEncarcelado()) {
            if (!tieneSalvoconducto()) {
                perderSalvoconducto();
                encarcelar = false;
                Diario.getInstance().ocurreEvento("Jugador " + nombre + "se libra de carcel");
            } else {
                encarcelar = true;
            }
        }
        return encarcelar;
    }

    boolean enBancarrota() {
        boolean bancarrota = false;

        if (saldo == 0) {
            bancarrota = true;
        }

        return bancarrota;
    }

    boolean encarcelar(int numCasillaCarcel) {
        if (debeSerEncarcelado()) {
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.getInstance().ocurreEvento("Jugador " + nombre + "va a la carcel");
        }

        return encarcelado;
    }

    boolean existeLaPropiedad(int ip) {
        boolean existir = true;

        if (propiedades.get(ip) == null) {
            existir = false;
        }

        return existir;
    }

    int getCasasMax() {
        return CasasMax;
    }

    int getHotelesMax() {
        return HotelesMax;
    }

    int getCasasPorHotel() {
        return CasasPorHotel;
    }

    protected String getNombre() {
        return nombre;
    }

    int getNumCasillaActual() {
        return numCasillaActual;
    }

    float getPrecioLibertad() {
        return PrecioLibertad;
    }

    float getPremioPorSalida() {
        return PasoPorSalida;
    }

    ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }

    boolean getPuedeComprar() {
        return puedeComprar;
    }

    protected float getSaldo() {
        return saldo;
    }

    public boolean isEncarcelado() {
        return encarcelado;
    }

    boolean modificarSaldo(float cantidad) {
        saldo += cantidad;
        Diario.getInstance().ocurreEvento("Saldo incrementado en " + cantidad);
        return true;
    }

    boolean moverACasilla(int numCasilla) {
        boolean b = true;

        if (isEncarcelado()) 
            b = false;
        else {
            numCasillaActual = numCasilla;
            puedeComprar = false;
            Diario.getInstance().ocurreEvento("Jugador " + nombre + "se mueve a " + numCasillaActual);
        }

        return b;
    }
    
    boolean obtenerSalvoconducto(Sorpresa s) {
        boolean b = true;
        
        if(isEncarcelado())
            b = false;
        else
            salvoconducto = s;
        
        return b;
    }
    
    boolean paga(float cantidad) {
        return modificarSaldo(cantidad * -1);
    }
    
    boolean pagaAlquiler(float cantidad) {
        boolean b = true;
        
        if(isEncarcelado())
            b = false;
        else
            paga(cantidad);
        
        return b;
    }
    
    boolean pagaImpuesto(float cantidad) {
        boolean b = true;
        
        if(isEncarcelado())
            b = false;
        else
            paga(cantidad);
        
        return b;
    }
    
    boolean pasaPorSalid() {
        modificarSaldo(PasoPorSalida);
        Diario.getInstance().ocurreEvento("Jugador " + getNombre() + " ha pasado por salida");
        return true;
    }
    
    private void perderSalvoconducto() {
        salvoconducto.usada();
        salvoconducto = null;
    }
    
    boolean puedeComprarCasilla() {
        if(isEncarcelado())
            puedeComprar = false;
        else
            puedeComprar = true;
        
        return puedeComprar;
    }
    
    private boolean puedeSalirCarcelPagando() {
        boolean b = false;
        
        if(getSaldo() >= getPrecioLibertad())
            b = true;
        
        return b;
    }
    
    private boolean puedoEdificarCasa(TituloPropiedad propiedad) {
        boolean b = false;
        
        if(propiedad.getNumCasas() < getCasasMax())
            if (getSaldo() >= propiedad.getPrecioEdificar()) {
                b = true;
            }
        
        return b;
    }
    
    private boolean puedoEdificarHotel(TituloPropiedad propiedad) {
        boolean b = false;
        
        if(propiedad.getNumHoteles() < getHotelesMax())
            if(propiedad.getNumCasas() == getCasasPorHotel())
                if (getSaldo() >= propiedad.getPrecioEdificar()) {
                    b = true;
                }
        
        return b;
    }
    
    private boolean puedoGastar(float precio) {
        boolean b = true;
        
        if(isEncarcelado())
            b = false;
        else if(precio >= getSaldo())
            b = true;
        else
            b = false;
        
        return b;
    }
    
    boolean recibe(float cantidad) {
        boolean b = true;
        
        if(isEncarcelado())
            b = false;
        else{
            modificarSaldo(cantidad);
            Diario.getInstance().ocurreEvento("Jugador " + getNombre() + " recibe " + cantidad);
        }
        
        return b;
    }
    
    boolean tieneAlgoQueGestionar() {
        boolean b = false;
        
        if(propiedades.size() > 0)
            b = true;
        
        return b;
    }
    
    boolean tieneSalvoconducto() {
        boolean b = false;
        
        if(salvoconducto != null)
            b = true;
        
        return b;
    }
    
    boolean vender(int ip) {
        boolean b = true;
        
        if(isEncarcelado())
            b = false;
        else{
            
            if(existeLaPropiedad(ip))
                if(propiedades.get(ip).vender(this)){
                    propiedades.remove(ip);
                    Diario.getInstance().ocurreEvento("Jugador " + getNombre() + " vende " + propiedades.get(ip).getNombre());
                    b = true;
                }
        }
        
        return b;
    }
}

File: Ruby-Civitas/jugador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class Jugador
    
    attr_accessor :nombre, :CASAS_MAX, :HOTELES_MAX, :CASAS_POR_HOTEL, :numCasillaActual, :PRECIO_POR_LIBERTAL, 
      :PASO_POR_SALIDA, :propiedades, :puedeComprar, :saldo, :encarcelado
    
    CASAS_MAX = 4
    HOTELES_MAX = 4
    SALDO_INICIAL = 7500
    CASAS_POR_HOTEL = 4
    PASO_POR_SALIDA = 1000
    PRECIO_POR_LIBERTAD = 200
    
    def initialize(nomb, jug = nil)
      if(jug != nil)
        @nombre = nomb
        @propiedades = []
        @encarcelado = false
        @numCasillaActual = 0
        @puedeComprar = true
        @saldo = SaldoInicial
        @salvoconducto = Sorpresa.new  
      else
        @propiedades = jug.propiedades[0..jug.propiedades.size-1]
        @nombre = jug.nombre
        @encarcelado = jug.encarcelado
        @numCasillaActual = jug.numCasillaActual
        @puedeComprar = jug.puedeComprar
        @saldo = jug.saldo
      end
    end
    
    def cantidadCasasHoteles()
      return @propiedades.size;
    end
    
    def <=>(jug)
      return @saldo <=> jug.saldo
    end
    
    def debeSerEncarcelado()
      encarcelar = false
      
      if(!@encarcelado)
        if(!tieneSalvoconducto())
          perderSalvoconducto()
          encarcelar = false
          Diario.instance.ocurre_evento("Jugador " + @nombre.to_s + "se libra de carcel")
        else
          encarcelar = true
        end
      end
      
      return encarcelar
    end
    
    def enBancarrota()
      bancarrota = false
      
      if(@saldo == 0)
        bancarrota = true
      end
      
      return bancarrota
    end
    
    def encarcelar(numCasillaCarcel)
      if(debeSerEncarcelado())
        moverAcasilla(numCasillaCarcel)
        @encarcelado = true
        Diario.instance.ocurre_evento("Jugador " + @nombre.to_s + "va a la carcel")
      end
      
      return @encarcelado
    end
    
    def existeLaPropiedad(ip)
      existir = true
      
      if(@propiedades[ip] == nil)
        existir = false
      end
      
      return existir
    end
    
    def modificarSaldo(cantidad)
      @saldo += cantidad
      Diario.instance.ocurre_evento("Saldo incrementado en " + cantidad.to_s)
      return true
    end
    
    def moverACasilla(numCasilla)
      b = true
      
      if(@encarcelado)
        b = false
      else
        @numCasillaActual = numCasilla
        @puedeComprar = false
        Diario.instance.ocurre_evento("Jugador " + @nombre.to_s + "se mueve a " + @numCasillaActual.to_s)
      end
      
      return b
    end
    
    def obtenerSalvoconducto(s)
      b = true
      
      if(@encarcelado)
        b = false
      else
        @salvoconducto = s
      end
      
      return b
    end
    
    def paga(cantidad)
      return modificarSaldo(cantidad * -1)
    end
    
    def pagaAlquiler(cantidad)
      b = true
      
      if(@encarcelado)
        b = false
      else
        paga(cantidad)
      end
      
      return b
    end
    
    def pagaImpuesto(cantidad)
      b = true
      
      if(@encarcelado)
        b = false
      else
        paga(cantidad)
      end
      
      return b
    end
    
    def pasaPorSalida()
      modificarSaldo(PASO_POR_SALIDA)
      Diario.instance.ocurre_evento("Jugador " + @nombre.to_s + "ha pasado por salida")
      return true
    end
    
    def perderSalvoConducto
      @salvoconducto.usada()
      @salvoconducto = nil
    end
    
    def puedeComprarCasilla()
      if(@encarcelado)
        @puedeComprar = false
      else
        @puedeComprar = true
      end
      
      return @puedeComprar
    end
    
    def puedeSalirCarcelPagando()
      b = false
      
      if(@saldo >= PRECIO_POR_LIBERTAD)
        b = true
      end
      
      return b
    end
    
    def puedeEdificarCasa(propiedad)
      b = false
      
      if(propiedad.numCasas < CASAS_MAX)
        if(@saldo >= propiedad.precioEdificar)
          b = true
        end
      end
      
      return b
    end
    
    def puedeEdificarHotel(propiedad)
      b = false
      
      if(propiedad.numHoteles < HOTELES_MAX)
        if(propiedad.numCasas == CASAS_POR_HOTEL)
          if(@saldo >= propiedad.precioEdificar)
            b = true
          end
        end
      end
      
      return b
    end
    
    def puedoGastar(precio)
      b = false
      
      if(@encarcelado)
        b = false
      else
        if(precio <= @saldo)
          b = true
        else
          b = false
        end
      end
      
      return b
    end
    
    def recibe(cantidad)
      b = true
      
      if(@encarcelado)
        b = false
      else
        modificarSaldo(cantidad)
        Diario.instance.ocurre_evento("Jugador " + @nombre.to_s + " recibe " + cantida.to_s)
      end
      
      return b
    end
    
    def tieneAlgoQueGestionar()
      b = false
      
      if(@propiedades.size > 0)
        b = true
      end
      
      return b
    end
    
    def tieneSalvoconducto()
      b = false
      
      if(@salvoconducto != nil)
        b = true
      end
      
      return b
    end
    
    def vender(ip)
      b = false
      
      if(@encarcelado)
        b = false
      else
        if(existeLaPropiedad(ip))
          if(@propiedades[ip].vender(this))
            @propiedades.delete_at(ip)
            Diario.instance.ocurre_evento("Jugador " + @nombre + " vende " + @propiedades[ip].nombre)
            b = true
          end
        end
      end
      
      return b
    end
    
    def toString()
      return "Jugador: " + @nombre + "\n"
      + "Encarcelado?: " + @encarcelado + "\n"
      + "Saldo: " + @saldo + "\n"
      + "Casilla Actual: " + @numCasillaActual
    end
  end
end
